tuna.go

```diff
@@ -1225,50 +1226,77 @@ func UpdateMetadata(
 	if subscriptionDuration > 3 {
 		subInterval = time.Duration(subscriptionDuration-3) * config.ConsensusDuration
 	}
-	nextSub := time.After(0)
+	var nextSub <-chan time.Time
 
 	go func() {
-		func() {
-			sub, err := client.GetSubscription(topic, address.MakeAddressString(client.PubKey(), identifier))
-			if err != nil {
-				log.Println("Get existing subscription error:", err)
-				return
-			}
+		for {
+			nextSub = time.After(0)
 
-			if len(sub.Meta) == 0 && sub.ExpiresAt == 0 {
-				return
-			}
+			func() {
+				sub, err := client.GetSubscription(topic, address.MakeAddressString(client.PubKey(), identifier))
+				if err != nil {
+					log.Println("Get existing subscription error:", err)
+					return
+				}
+
+				if len(sub.Meta) == 0 && sub.ExpiresAt == 0 {
+					return
+				}
+
+				if sub.Meta != string(metadataRaw) {
+					log.Println("Existing subscription meta need update.")
+					return
+				}
+
+				height, err := client.GetHeight()
+				if err != nil {
+					log.Println("Get current height error:", err)
+					return
+				}
 
-			if sub.Meta != string(metadataRaw) {
-				log.Println("Existing subscription meta need update.")
+				if sub.ExpiresAt-height < 3 {
+					log.Println("Existing subscription is expiring")
+					return
+				}
+
+				log.Println("Existing subscription expires after", sub.ExpiresAt-height, "blocks")
+
+				maxSubDuration := float64(sub.ExpiresAt-height) * float64(config.ConsensusDuration)
+				nextSub = time.After(time.Duration((1 - rand.Float64()*subscribeDurationRandomFactor) * maxSubDuration))
+			}()
+
+			select {
+			case <-nextSub:
+			case <-closeChan:
 				return
 			}
 
-			height, err := client.GetHeight()
+			subFee, err := common.StringToFixed64(subscriptionFee)
 			if err != nil {
-				log.Println("Get current height error:", err)
-				return
+				log.Println("Parse subscription fee error:", err)
 			}
 
-			if sub.ExpiresAt-height < 3 {
-				log.Println("Existing subscription is expiring")
-				return
+			if subFee > 0 {
+				balance, err := client.Balance()
+				if err != nil {
+					log.Println("Get balance error:", err)
+				} else {
+					if subFee > balance.ToFixed64() {
+						subFee = balance.ToFixed64()
+					}
+				}
 			}
 
-			log.Println("Existing subscription expires after", sub.ExpiresAt-height, "blocks")
+			addToSubscribeQueue(client, identifier, topic, int(subscriptionDuration), string(metadataRaw), &nkn.TransactionConfig{Fee: subFee.String()}, subscriptionReplaceTxPool)
 
-			maxSubDuration := float64(sub.ExpiresAt-height) * float64(config.ConsensusDuration)
-			nextSub = time.After(time.Duration((1 - rand.Float64()*subscribeDurationRandomFactor) * maxSubDuration))
-		}()
+			nextSub = time.After(time.Duration((1 - rand.Float64()*subscribeDurationRandomFactor) * float64(subInterval)))
 
-		for {
 			select {
 			case <-nextSub:
```

Comment:
Can we keep only `maxCheckSubscribeInterval` as check interval?

Reply:
We can't because `subInterval` is configurable and can be less than `maxCheckSubscribeInterval`